### PR TITLE
Update sourceSetOutput example to use Provider

### PR DESCRIPTION
**Pull request by bigdaz**
The "example how to work with generated resources" was using an old 'builtBy'
parameter to wire task dependencies. This change updates the example to use
a `DirectoryProperty` output and `task.flatMap` to do the wiring.

### CODE REVIEW

File: subprojects/plugins/src/main/java/org/gradle/api/tasks/SourceSetOutput.java [hunk 1]
```diff
@@ -54,33 +54,30 @@
  *
  * <pre class='autoTested'>
  * plugins {
- *     id 'java'
+ *   id 'java'
  * }
  *
- * def generatedResources = "$buildDir/generated-resources/main"
+ * def generateResourcesTask = tasks.register("generate-resources", GenerateResourcesTask) {
+ *   resourcesDir.set(file("$buildDir/generated-resources/main"))
```
Comment (big-guy) [inline subprojects/plugins/src/main/java/org/gradle/api/tasks/SourceSetOutput.java:61]: can you change this around to use `layout.buildDirectory.dir(...)`?

File: subprojects/plugins/src/main/java/org/gradle/api/tasks/SourceSetOutput.java [hunk 2]
```diff
@@ -54,33 +54,30 @@
...
+ * }
  *
  * sourceSets {
  *   main {
- *     //let's register an output folder on the main SourceSet:
- *     output.dir(generatedResources, builtBy: 'generateMyResources')
- *     //it is now a part of the 'main' classpath and will be a part of the jar
+ *     // Register the generated 'resourcesDir' as an output of the main source set, retaining the task dependency.
+ *     output.dir(generateResourcesTask.flatMap {it.resourcesDir} )
```
Comment (big-guy) [inline subprojects/plugins/src/main/java/org/gradle/api/tasks/SourceSetOutput.java:67]: Does this work if you just leave it as `generateResourcesTask`?

File: subprojects/plugins/src/main/java/org/gradle/api/tasks/SourceSetOutput.java [hunk 3]
```diff
@@ -54,33 +54,30 @@
...
  *   }
  * }
  *
- * //a task that generates the resources:
- * task generateMyResources {
- *   outputs.dir generatedResources
- *   doLast {
- *     def generated = new File(generatedResources, "myGeneratedResource.properties")
+ * abstract class GenerateResourcesTask extends DefaultTask {
```
Comment (big-guy) [inline subprojects/plugins/src/main/java/org/gradle/api/tasks/SourceSetOutput.java:71]: Very  nice!

**Review by big-guy — CHANGES_REQUESTED** [review]

File: subprojects/plugins/src/main/java/org/gradle/api/tasks/SourceSetOutput.java (hunk 2 above)
Comment (bigdaz) [reply subprojects/plugins/src/main/java/org/gradle/api/tasks/SourceSetOutput.java]: Yes it does. Updated.

File: subprojects/plugins/src/main/java/org/gradle/api/tasks/SourceSetOutput.java (hunk 1 above)
Comment (bigdaz) [reply subprojects/plugins/src/main/java/org/gradle/api/tasks/SourceSetOutput.java]: Done

**Comment by big-guy** [comment]
@bot-gradle test and merge

**Pull request merged**